### DISCUSSION TITLE
How to create a procedural model or mesh in code?

### DISCUSSION THREAD

**Discussion by tebjan** [stride3d/stride#1041] [Q&A] [answered]
This question has extensive answers on Stackexchange: [How to create a mesh in code?](https://gamedev.stackexchange.com/questions/182379/how-to-create-a-mesh-in-code)

The question:

Getting my feet wet with Stride (formerly Xenko), I want to build a level editor where a user can drag the mouse to create walls at runtime.
I found documentation on how to [immediately draw vertices](https://doc.stride3d.net/latest/en/manual/graphics/low-level-api/draw-vertices.html) per frame, but I want those vertices to be persistently stored for a mesh.

Does anyone have a link or can explain to me which steps are needed to do this?

**Comment by tebjan** [comment, marked as answer]
The accepted answer is:

You can make your own procedural model by inheriting from `PrimitiveProceduralModelBase`. This has several advantages:
* Write less code
* Available in Game Studio with preview and gizmo in scene editor
* Properties (including material) are editable in Game Studio
* Auto-generate tangents and bi-tangents required for normal maps
* Auto-generate bounding box and bounding sphere for correct shadow map bias
* Auto-generate additional texture coordinate channels
* Model space position offset
* Better upload/render performance due to correct data types

![image](https://user-images.githubusercontent.com/1094716/113584573-593f1a80-962b-11eb-91a0-7e71b1032bfc.png)

The workflow to add the model in Game Studio is then:
* Add any procedural model (Cube, Plane, etc.)
* Change the dropdown to your custom procedural model implementation
* The workflow in code or script is:

Add an instance of the script below to any entity
Full sources - including model update per frame - are available here: https://github.com/tebjan/StrideProceduralModel

Code of the custom model:

```csharp
[DataContract("MyProceduralModel")]
[Display("MyModel")] // This name shows up in the procedural model dropdown list
public class MyProceduralModel : PrimitiveProceduralModelBase
{
    // A custom property that shows up in Game Studio
    /// <summary>
    /// Gets or sets the size of the model.
    /// </summary>
    public Vector3 Size { get; set; } = Vector3.One;

    protected override GeometricMeshData<VertexPositionNormalTexture> CreatePrimitiveMeshData()
    {
        // First generate the arrays for vertices and indices with the correct size
        var vertexCount = 4;
        var indexCount = 6;
        var vertices = new VertexPositionNormalTexture[vertexCount];
        var indices = new int[indexCount];

        // Create custom vertices, in this case just a quad facing in Y direction
        var normal = Vector3.UnitZ;
        vertices[0] = new VertexPositionNormalTexture(new Vector3(-0.5f, 0.5f, 0) * Size, normal, new Vector2(0, 0));
        vertices[1] = new VertexPositionNormalTexture(new Vector3(0.5f, 0.5f, 0) * Size, normal, new Vector2(1, 0));
        vertices[2] = new VertexPositionNormalTexture(new Vector3(-0.5f, -0.5f, 0) * Size, normal, new Vector2(0, 1));
        vertices[3] = new VertexPositionNormalTexture(new Vector3(0.5f, -0.5f, 0) * Size, normal, new Vector2(1, 1));

        // Create custom indices
        indices[0] = 0;
        indices[1] = 1;
        indices[2] = 2;
        indices[3] = 1;
        indices[4] = 3;
        indices[5] = 2;

        // Create the primitive object for further processing by the base class
        return new GeometricMeshData<VertexPositionNormalTexture>(vertices, indices, isLeftHanded: false) { Name = "MyModel" };
    }
}
```
Code of the script that adds the model by code:

```csharp
public class AddMyModelScript : AsyncScript
{
    // Declared public member fields and properties will show in the game studio
    public float RotationSpeed { get; set; } = 1;

    public override async Task Execute()
    {
        // Setup the custom model
        await CreateMyModel();

        while (Game.IsRunning)
        {
            // Do stuff every new frame
            Entity.Transform.Rotation *= Quaternion.RotationY(MathUtil.DegreesToRadians(RotationSpeed));

            await Script.NextFrame();
        }
    }

    async Task CreateMyModel()
    {
        // The model classes
        var myModel = new MyProceduralModel();
        var model = new Model();
        var modelComponent = new ModelComponent(model);

        // Generate the procedual model
        myModel.Generate(Services, model);

        // Add a meterial
        var material = Content.Load<Material>("MyModel Material");
        model.Add(material);

        // Add everything to the entity
        Entity.Add(modelComponent);
    }
}
```

**Reply by tagcode** [reply]
>Available in Game Studio with preview and gizmo in scene editor

I've created a procedural model class using your example as template. I can instantiate and parametrize it as a model by creating a .sdpromodel yaml-file manually. 

However, my procedural model class is not visible to the Game Studio, only the built-in ones are.

Is there anything I do to make Game Studio discover them?
<img width="513" alt="gamestudio" src="https://user-images.githubusercontent.com/5236113/211911112-66d2b94c-3797-44cb-9cdd-c3a49318cb07.png">

**Reply by manio143** [reply]
Yes, but at the moment it would mean taking dependencies in your project on libraries which are only used by the editor. This kind of work would fall under a custom asset/editor plugin and there's an ongoing issue around introducing plugins but it won't get completed soon.

**Reply by tagcode** [reply]
Understod, thanks for info. I opened an [issue](https://github.com/stride3d/stride/issues/1583).

**Comment by jasonswearingen** [comment]
additional reading:
- https://github.com/profan/XenkoByteSized
- https://github.com/tebjan/StrideProceduralModel
- https://github.com/jeske/LoadObjTest
- https://doc.stride3d.net/4.0/en/manual/scripts/create-a-model-from-code.html
- C#BeginnerSample --> LoadingContentDemo

**Comment by VaclavElias** [comment]
Also recently added which might simplify creation of such objects

https://stride3d.github.io/stride-community-toolkit/manual/code-only/examples/procedural-geometry.html
